[AArch64][TargetParser] Split FMV and extensions

tommat01 opened the pull request:
FMV extensions are really just mappings from FMV feature names to lists of backend features for codegen. Split them out into their own separate file.

tmatheson-arm commented on llvm/lib/Target/AArch64/AArch64FMV.td:
```diff
@@ -0,0 +1,84 @@
+
+    // A C++ expression giving the number of the bit in the FMV ABI.
+    // Currently this is given as a value from the enum "CPUFeatures".
+    // If this is not set, it indicates that this is not an FMV extension.
```
remove

labrinea commented on llvm/lib/Target/AArch64/AArch64FMV.td:
```diff
@@ -0,0 +1,84 @@
+    string BackendFeatures = f;
+
+    // The FMV priority.
+    int Priority = p;
```
unsigned?

labrinea commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -13664,16 +13665,23 @@ QualType ASTContext::getCorrespondingSignedFixedPointType(QualType Ty) const {
 
+// Given a list of FMV features, add each of their backend features to the list.
+static void
+GetFMVBackendFeaturesFor(const llvm::SmallVector<StringRef, 8> FMVFeatStrings,
```
pass by reference?

labrinea commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -13754,11 +13750,14 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+    TV->getFeatures(Feats);
+    std::vector<std::string> Features;
+    GetFMVBackendFeaturesFor(Feats, Features);
     Feats.insert(Feats.begin(),
```
s/Feats/Features

tommat01 committed:
[AArch64][TargetParser] Split FMV and extensions

tommat01 committed:
clang-format

labrinea commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -13683,9 +13683,9 @@ static std::vector<std::string> getFMVBackendFeaturesFor(
+    if (auto FMVExt = llvm::AArch64::parseFMVExtension(F)) {
       SmallVector<StringRef, 8> Feats;
-      FMVExt->DependentFeatures.split(Feats, ',', -1, false);
+      FMVExt->Features.split(Feats, ',', -1, false);
```
You can use getImpliedFeatures() here.

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into split_fmv_and_extensions

tommat01 committed:
use FMVExt->getImpliedFeatures()

tommat01 committed:
Remove extensions for BTI, WfXT, and DIT.

These were exposed out of necessity due to the limitations of the
previous implementation. Revert them back to being internal backend
features.

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into split_fmv_and_extensions

labrinea reviewed (approved):
Looking good Tomas and thanks for your effort to tidy things up. I know this is NFC, but can we rectify the implied features for rcpc2 and rcpc3 by adding rcpc-immo as shown here https://github.com/llvm/llvm-project/pull/92319/files#diff-0df0a19b911b768165642047ad84892b945f9e2e652f91e8d9ed911b21915ec4R488 ?

tmatheson-arm commented:
> Looking good Tomas and thanks for your effort to tidy things up. I know this is NFC, but can we rectify the implied features for rcpc2 and rcpc3 by adding rcpc-immo as shown here https://github.com/llvm/llvm-project/pull/92319/files#diff-0df0a19b911b768165642047ad84892b945f9e2e652f91e8d9ed911b21915ec4R488 ?

That would be better as a separate PR.
